### PR TITLE
Overhaul Observation Transformation API + Better 360 and multisensor support.

### PR DESCRIPTION
**Pull request by Skylion007**
## Motivation and Context

Solves #367 

Okay, this ObservationTransformer API does the following:
* Adds a new observation transformer registry
* Videos generated now show the final output of the network after all the observation transformers have been fed in.
* Still not ideal since Observations may need to be normalized/ denormalized in each operation. Should fix this longterm.
* Implemented a demo cubemap2equirect that stitches together 6 sensors into one using this API. Creating a 360 Sensor.
* Remove unnecessary float conversion in PPO trainer.
* Reads in video input from batched input now.
* ObservationTransformers separated from Policies.
* Introduces a new API to run.py that can feed in a config without writing it to disk.
* Add more documentations
* Add more typing to better document what is going on.
* Adds tests to cover the change.
* Changes the policy fromConfig API to not have access to the full environment.

## How Has This Been Tested
With Pytest and by splitting the ResizeCenterCropper into two PyTorch modules

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: habitat_baselines/common/obs_transformers.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+
+@baseline_registry.register_obs_transformer()
+class ResizeShortestEdge(ObservationTransformer):
+    def __init__(self, size: int, channels_last: bool = False):
+        r"""An nn module the resizes your the shortest edge of the input while maintaining aspect ratio.
+        This module assumes that all images in the batch are of the same size.
+        Args:
+            size: The size you want to resize the shortest edge to
+            channels_list: indicates if channels is the last dimension
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:51]:
```suggestion
            channels_last: indicates if channels is the last dimension
```

File: habitat_baselines/common/obs_transformers.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        super().__init__()
+        self._size = size
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:54]: You can try using `@attr.s(auto_attribs=True, kw_only=True)` to avoid this code.

File: habitat_baselines/common/obs_transformers.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        self.channels_last = channels_last
+
+    def transform_observation_space(
+        self,
+        observation_space: SpaceDict,
+        trans_keys=("rgb", "depth", "semantic"),
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:60]: Makes more sense to have list here.
```suggestion
        trans_keys: List[str]=["rgb", "depth", "semantic"],
```

File: habitat_baselines/common/obs_transformers.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    ):
+        size = self._size
+        observation_space = copy.deepcopy(observation_space)
+        if size:
+            for key in observation_space.spaces:
+                if key in trans_keys:
+                    # In the observation space dict, the channels are always last
+                    h, w = get_image_height_width(
+                        observation_space.spaces[key], channels_last=True
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:69]:
```suggestion
                        observation_space.spaces[key], channels_last=self.channels_last
```

File: habitat_baselines/common/obs_transformers.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                    )
+                    logger.info(
+                        "Center cropping observation size of %s from %s to %s"
+                        % (key, (h, w), size)
+                    )
+
+                    observation_space.spaces[key] = overwrite_gym_box_shape(
+                        observation_space.spaces[key], size
+                    )
+        self.observation_space = observation_space
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:140]: Looks like never used.
```suggestion
```

File: habitat_baselines/common/obs_transformers.py (hunk 3 above)
Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:60]: Should be a tuple per flake8 not a list.

File: habitat_baselines/common/obs_transformers.py (hunk 4 above)
Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:69]: Nope, observations are always channels last whether the the transformation is or not.

File: habitat_baselines/common/obs_transformers.py (hunk 2 above)
Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:54]: @mathfac How do I so while preserving the docstring though?

Comment (erikwijmans) [reply habitat_baselines/common/obs_transformers.py:54]: The docstring for the constructor is generally put on the class instead of `__init__`

File: habitat_baselines/config/default.py [hunk 6]
```diff
@@ -59,6 +59,16 @@
 _C.RL.POLICY = CN()
 _C.RL.POLICY.name = "PointNavBaselinePolicy"
 # -----------------------------------------------------------------------------
+# OBS_TRANSFORMS CONFIG
+# -----------------------------------------------------------------------------
+_C.RL.POLICY.OBS_TRANSFORMS = CN()
+_C.RL.POLICY.OBS_TRANSFORMS.active = tuple()
```
Comment (mathfac) [inline habitat_baselines/config/default.py:65]: Maybe more explicit and conventional:
```suggestion
_C.RL.POLICY.OBS_TRANSFORMS.ENABLED_TRANSFORMS = tuple()
```

**Review by mathfac — APPROVED** [review]
Checked the logic again. Looks more extendable now! Minor comment.  
Thank you for bringing typing to policies code.

File: habitat_baselines/common/obs_transformers.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                    for sensor in self.trans_keys
+                    if sensor in observations
+                }
+            )
+        return observations
+
+    @classmethod
+    def from_config(cls, config: Config, envs):
+        return cls(
+            (
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:180]:
```suggestion
           size = (
```

File: habitat_baselines/common/obs_transformers.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            )
+        )
+
+
+def get_active_obs_transforms(
+    config: Config, envs
+) -> List[ObservationTransformer]:
+    active_obs_transforms = []
+    if hasattr(config.RL.POLICY, "OBS_TRANSFORMS"):
+        obs_transform_names = config.RL.POLICY.OBS_TRANSFORMS.active
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:192]:
```suggestion
        obs_transform_names = config.RL.POLICY.OBS_TRANSFORMS.ENABLED_TRANSFORMS
```

File: habitat_baselines/common/obs_transformers.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:5]:
```suggestion

# LICENSE file in the root directory of this source tree.
```

**Comment by mathfac** [comment]
Do we have tests covering these changes? Thank you

File: habitat_baselines/config/default.py (hunk 6 above)
Comment (Skylion007) [reply habitat_baselines/config/default.py:65]: @mathfac Should we update POLICY as well? I just wanted to be consistent with our .name and stuff.

File: habitat_baselines/common/utils.py [hunk 10]
```diff
@@ -148,11 +99,7 @@ def batch_obs(
             batch[sensor].append(_to_tensor(obs[sensor]))
 
     for sensor in batch:
-        batch[sensor] = (
```
Comment (mathfac) [inline habitat_baselines/common/utils.py:151]: cc @erikwijmans, is this change safe?

File: habitat_baselines/rl/ddppo/algo/ddppo_trainer.py [hunk 11]
```diff
@@ -188,8 +201,10 @@ def train(self) -> None:
 
         observations = self.envs.reset()
         batch = batch_obs(observations, device=self.device)
+        batch = apply_obs_transforms_batch(batch, self.obs_transforms)
 
-        obs_space = self.envs.observation_spaces[0]
+        # obs_space = self.envs.observation_spaces[0]
+        obs_space = self.obs_space
```
Comment (mathfac) [inline habitat_baselines/rl/ddppo/algo/ddppo_trainer.py:207]: Maybe, Remove both lines, and use `self.obs_space` instead of `obs_space`

File: habitat_baselines/rl/ppo/ppo_trainer.py [hunk 12]
```diff
@@ -60,7 +65,16 @@ def _setup_actor_critic_agent(self, ppo_cfg: Config) -> None:
         logger.add_filehandler(self.config.LOG_FILE)
 
         policy = baseline_registry.get_policy(self.config.RL.POLICY.name)
-        self.actor_critic = policy.from_config(self.config, self.envs)
+        observation_space = self.envs.observation_spaces[0]
+        self.obs_transforms = get_active_obs_transforms(self.config)
+        observation_space = apply_obs_transforms_obs_space(
+            observation_space, self.obs_transforms
+        )
+        # for active_obs in self.obs_transforms:
```
Comment (mathfac) [inline habitat_baselines/rl/ppo/ppo_trainer.py:73]: Comments left.

File: habitat_baselines/common/obs_transformers.py [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+    def from_config(cls, config: Config):
+        return cls(
+            (
+                config.RL.POLICY.OBS_TRANSFORMS.CENTER_CROPPER.HEIGHT,
+                config.RL.POLICY.OBS_TRANSFORMS.CENTER_CROPPER.WIDTH,
+            )
+        )
+
+
+class Cube2Equirec(nn.Module):
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:193]: Need tests for this functionality and let's move to separate PR, as soon as this PR will be ready.

File: habitat_baselines/rl/ddppo/algo/ddppo_trainer.py (hunk 11 above)
Comment (Skylion007) [reply habitat_baselines/rl/ddppo/algo/ddppo_trainer.py:207]: This mutates the observation space in a very specific way though and the original code did not modify the observation space return from the envs so I am not sure that is the proper move.

File: habitat_baselines/common/utils.py (hunk 10 above)
Comment (Skylion007) [reply habitat_baselines/common/utils.py:151]: @mathfac All tests pass so presumably and any RGB input is already converted implicitly to float in the encoders.

File: habitat_baselines/common/obs_transformers.py [hunk 14]
```diff
@@ -0,0 +1,630 @@
...
+    def __init__(self, equ_h, equ_w, cube_length, CUDA=False):
```
Comment (erikwijmans) [inline habitat_baselines/common/obs_transformers.py:216]: CUDA really shouldn't be an arg to `__init__` IMO.  You can do 
```
self.grid  = self.grid.to(batch.device)
self.orientation_mask = self.orientation_mask.to(batch.device)
```
in forward and that will be a no-op after the first call.

File: habitat_baselines/common/utils.py (hunk 10 above)
Comment (erikwijmans) [reply habitat_baselines/common/utils.py:151]: Yeah, it should be fine

**Comment by Skylion007** [comment]
All tests are now passing. Ready to get this merged. Only thing I am unsure of is the ObsTransformer name and if I should move it from the RL.POLICY to just RL.

File: habitat_baselines/common/obs_transformers.py [hunk 15]
```diff
@@ -0,0 +1,623 @@
...
+        return observations
+
+    @classmethod
+    def from_config(cls, config: Config):
+        return cls(config.RL.POLICY.OBS_TRANSFORMS.RESIZE_SHORTEST_EDGE.SIZE)
+
+
+@baseline_registry.register_obs_transformer()
+class CenterCropper(ObservationTransformer):
+    def __init__(
```
Comment (mathfac) [inline habitat_baselines/common/obs_transformers.py:137]: We need doc string.

**Comment by Skylion007** [comment]
I am going to move the OBS_TRANSFORMER out of RL.POLICY to RL. and willl add the doc strings.

**Comment by Skylion007** [comment]
I am still not sure what scope to put the config under, thoughts @erikwijmans I could RL.OBS_TRANSFORMER, but I could also see the arguement for RL.POLICY.OBS_TRANSFORMER. It's general enough to just be at the root node too though.

**Review by mathfac — APPROVED** [review]
